File: python/sorting/heap_sort_production.py
from collections.abc import Callable, MutableSequence
from typing import Any, TypeVar

T = TypeVar("T")


def heap_sort_in_place(
    values: MutableSequence[T],
    *,
    key: Callable[[T], Any] | None = None,
    reverse: bool = False,
) -> MutableSequence[T]:
    key_fn = key or (lambda item: item)
    length = len(values)

    def before(left: T, right: T) -> bool:
        left_key = key_fn(left)
        right_key = key_fn(right)
        return left_key > right_key if reverse else left_key < right_key

    def sift_down(start: int, end: int) -> None:
        root = start

        while True:
            child = root * 2 + 1

            if child >= end:
                break

            right = child + 1

            if right < end and before(values[child], values[right]):
                child = right

            if not before(values[root], values[child]):
                break

            values[root], values[child] = values[child], values[root]
            root = child 

    for start in range(length // 2 - 1, -1, -1):
        sift_down(start, length)

    for end in range(length - 1, 0, -1):
        values[0], values[end] = values[end], values[0]
        sift_down(0, end)

    return values

File: java/sorting/HeapSortProduction.java
import java.util.Comparator;
import java.util.List;
import java.util.Objects;

public final class HeapSortProduction {

    private HeapSortProduction() {
    }

    public static <T extends Comparable<? super T>> void heapSortInPlace(List<T> values) {
        heapSortInPlace(values, Comparator.naturalOrder());
    }

    public static <T> void heapSortInPlace(List<T> values, Comparator<? super T> comparator) {
        Objects.requireNonNull(values, "values");
        Objects.requireNonNull(comparator, "comparator");

        int length = values.size();

        for (int start = length / 2 - 1; start >= 0; start--) {
            siftDown(values, start, length, comparator);
        }

        for (int end = length - 1; end > 0; end--) {
            swap(values, 0, end);
            siftDown(values, 0, end, comparator);
        }
    }

    private static <T> void siftDown(List<T> values, int start, int end, Comparator<? super T> comparator) {
        int root = start;

        while (true) {
            int child = root * 2 + 1;

            if (child >= end) {
                break;
            }

            int right = child + 1;

            if (right < end && comparator.compare(values.get(child), values.get(right)) < 0) {
                child = right;
            }

            if (comparator.compare(values.get(root), values.get(child)) >= 0) {
                break;
            }

            swap(values, root, child);
            root = child;
        }
    }

    private static <T> void swap(List<T> values, int left, int right) {
        T value = values.get(left);
        values.set(left, values.get(right));
        values.set(right, value);
    }
}
